### ISSUE TITLE
Support callback path based on IApplicationBuilder.Map paths

### ISSUE THREAD

**Issue by span** [ActiveLogin/ActiveLogin.Authentication#123]
**Is your feature request related to a problem? Please describe.**
The current callback path does not work together with path matches defined in `IApplicationBuilder.Map`. If adding the path match prefix from `IApplicationBuilder.Map` to the callback path, the callback endpoint becomes nested.

**What area is it related to**
BankIdAuthenticationHandler

**Describe the solution you'd like**
An option to add the path that I use in `IApplicationBuilder.Map` to the callback path. This can then be used in an override of `ShouldHandleRequestAsync` to handle the nested path.

**Describe alternatives you've considered**
URL rewriting etc.

**Additional context**
I have an open stack overflow question on how to work around the issue but it does not seem possible without changing the `BankIdAuthenticationHandler`.

https://stackoverflow.com/questions/58249177/using-remoteauthenticationhandler-callbackpath-with-iapplicationbuilder-path-mat?noredirect=1#comment102983312_58249177

I could spend the time implementing this if you think it it necessary and will make for a good addition.

**Comment by fredrik-lundin** [comment]
Hi,

Thanks for posting such a detailed and well explained issue. I definitely think your scenario is something that we should support. Even though it might not be the most common use-case to wrap whole routing inside a `map`, it's certainly not an extreme edge case in my eyes.

The answer you've got in your stack overflow thread looks promising to me (https://stackoverflow.com/a/58317651/2521893). It points out that the _matched path segments_ is added to `HttpRequest.PathBase` when running inside the context of a `map` . 

From the asp.net core docs (https://docs.microsoft.com/en-us/aspnet/core/fundamentals/middleware/?view=aspnetcore-3.0#use-run-and-map):

> When Map is used, the matched path segments are removed from HttpRequest.Path and appended to HttpRequest.PathBase for each request.

So a possible solution could probably be to register a `IHttpContextAccessor`, then use that inside to `RemoteAuthenticationHandler` and then do something like what is suggested in the Stack overflow answer:

```
HttpContext context = _httpContextAccessor.HttpContext;
string returnUrl = UrlEncoder.Encode(context.Request.PathBase + Options.CallbackPath);
```

Nothing would make us happier than if you'd like to contribute with a pull request yourself! Either go with something like suggested above or if you have a better solution in mind feel free to try it out. I will gladly help out in discussion around this if needed!

**Comment by span** [comment]
Very nice, I will dig right into it.

I am not sure why you think I need the `IHttpContextAccessor` though. There is an `HttpContext` available in the base class `RemoteAuthenticationHandler` that it receives in the `InitializeAsync` method.

Is there any documentation that you know of that should be updated together with this change?

**Comment by fredrik-lundin** [comment]
> I am not sure why you think I need the `IHttpContextAccessor` though. There is an `HttpContext` available in the base class `RemoteAuthenticationHandler` that it receives in the `InitializeAsync` method.

Even better! 🙌🎉 Didn't know that from the top of my head 😳
 
> Is there any documentation that you know of that should be updated together with this change?

Not sure that is needed tbh. 

If I understand correctly (haven't had time to look into this closely) this change would be transparent to the end-user, wouldn't it? If you are inside of a `map`, then `PathBase` would be assigned the value of that path. If you're not inside a `map`, I would assume that the `PathBase` is just empty.

So the change would "just work" for everyone, without any extra configuration needed I guess. Please enlighten me if I'm not understanding this correctly 😄

**Comment by PeterOrneholm** [comment]
Hi @span!

Thanks for a great Issue, done "by the book" :)
Your solution seems good! It would be great, as @fredrik-lundin said, if you can write two tests. One verifying the setup with map and one without. We have one or two samples of test for this flow already.

Btw, not that we are joining "Hacktober" and you'll recevie a free t-shirt if you get your PR merged in October :) #124 

/Peter

**Comment by span** [comment]
Wow! I'm struggling a bit with the tests at the moment but hope to have a PR today.

**Comment by span** [comment]
Here is a first suggestion.

The implementation is simple enough. I just get the path base and add it to the login url and the callback url. Adding it to login url seems like the correct thing to do since they will probably match most of the time. 

The tests took some more work. I have added a test that sets up a pipeline with the path base. I could not figure out a way to do this without refactoring the app configuration for the tests first.

The default case, where path base is not defined, is already covered by the existing test `Challenge_Redirects_To_BankIdAuthentication_Login`.

I would love to hear your thoughts before submitting a PR.

**Comment by PeterOrneholm** [comment]
> I would love to hear your thoughts before submitting a PR.

Had a look at your commit now, looks great from what I can see! You can make a PR so we can see that all tests are green etc. Also easier to comment if we were to find anything before merging :)

Btw, GitHub has Draft PR:s nowdays. They are great for early feedback, but can't be merged until the author (you in this case) says it's ready to do so :)

**Comment by PeterOrneholm** [comment]
Ahh, my fault, I've tried upgrading to core 3. Will fix this. You still get your t-shirt :)

**Comment by span** [comment]
I think I've found a bug with this implementation in combination with the absolute URI resolving in `BankIdApiController`. 

In the controller the following method is used to create an absolute redirect URI:

```
private string GetAbsoluteUrl(string returnUrl)
        {
            var absoluteUri = $"{Request.Scheme}://{Request.Host.ToUriComponent()}{Request.PathBase.ToUriComponent()}";
            return absoluteUri + returnUrl;
        }
```

As you see, it uses the `PathBase` to create the url. This probably means that we will get nested paths in this redirect URI. I ran across it when working on #119 and trying to redirect the user back to login page after cancellation.

I think it will be enough to remove the pathbase from the method `GetAbsoluteUrl` since the path base is now being resolved in the `BankIdAuthenticationHandler`.

**Comment by PeterOrneholm** [comment]
Will close this once #128 is merged.

**Comment by PeterOrneholm** [comment]
> It seems the pipeline perhaps got bitten by this: [microsoft/azure-pipelines-tasks#11569](https://github.com/microsoft/azure-pipelines-tasks/issues/11569)

YTes, it was that one :=) I fixed it during the weekend. Thank you for the link!

**Comment by PeterOrneholm** [comment]
@span: With your PR (#128) now merged, would you agree this is now resolved so I can close it?

**Comment by span** [comment]
@PeterOrneholm yes 👯‍♂ 🎉 ☕️
